macros/provider.go

```diff
@@ -0,0 +1,155 @@
+package macros
+
+import (
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/prebid/openrtb/v19/openrtb2"
+	"github.com/prebid/prebid-server/exchange/entities"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+const (
+	MacroKeyBidID       = "PBS-BIDID"
+	MacroKeyAppBundle   = "PBS-APPBUNDLE"
+	MacroKeyDomain      = "PBS-DOMAIN"
+	MacroKeyPubDomain   = "PBS-PUBDOMAIN"
+	MacroKeyPageURL     = "PBS-PAGEURL"
+	MacroKeyAccountID   = "PBS-ACCOUNTID"
+	MacroKeyLmtTracking = "PBS-LIMITADTRACKING"
+	MacroKeyConsent     = "PBS-GDPRCONSENT"
+	MacroKeyBidder      = "PBS-BIDDER"
+	MacroKeyIntegration = "PBS-INTEGRATION"
+	MacroKeyVastCRTID   = "PBS-VASTCRTID"
+	MacroKeyTimestamp   = "PBS-TIMESTAMP"
+	MacroKeyAuctionID   = "PBS-AUCTIONID"
+	MacroKeyChannel     = "PBS-CHANNEL"
+	MacroKeyEventType   = "PBS-EVENTTYPE"
+	MacroKeyVastEvent   = "PBS-VASTEVENT"
+)
+const (
+	CustomMacroPrefix = "PBS-MACRO-"
+)
+
+var (
+	bidLevelKeys = []string{MacroKeyBidID, MacroKeyBidder, MacroKeyVastEvent, MacroKeyEventType, MacroKeyVastCRTID}
+)
+
+type MacroContext struct {
+	Bid            *entities.PbsOrtbBid
+	Imp            *openrtb2.Imp
+	Seat           string
+	VastCreativeID string
+	VastEventType  string
+	EventElement   string
+}
+
+type macroProvider struct {
+	// macros map stores macros key values
+	macros map[string]string
+}
+
+// NewBuilder returns the instance of macro buidler
+func NewProvider(reqWrapper *openrtb_ext.RequestWrapper) *macroProvider {
+	macroProvider := &macroProvider{macros: map[string]string{}}
+	macroProvider.populateRequestMacros(reqWrapper)
+	return macroProvider
+}
+
+func (b *macroProvider) populateRequestMacros(reqWrapper *openrtb_ext.RequestWrapper) {
+	b.macros[MacroKeyTimestamp] = strconv.Itoa(int(time.Now().Unix()))
+	reqExt, err := reqWrapper.GetRequestExt()
+	if err == nil && reqExt != nil && reqExt.GetPrebid() != nil {
+		for key, value := range reqExt.GetPrebid().Macros {
+			customMacroKey := CustomMacroPrefix + key       // Adding prefix PBS-MACRO to custom macro keys
+			b.macros[customMacroKey] = truncate(value, 100) // limit the custom macro value  to 100 chars only
```

Comment:
I'd recommend making the max length of a macro value a constant rather than hard coding it

Reply:
Done